src/toil_scripts/batch_alignment/bwa_alignment.py

```diff
@@ -1,223 +1,248 @@
-    Configuration file has one sample per line, with the following format:  UUID\t1st_url\t2nd_url
-
-    :param JobFunctionWrappingJob job: Passed by Toil automatically
-    :param dict shared_ids: stores FileStoreIDs
-    :param Namespace inputs: Input arguments
-    """
-    if inputs.sample:
-        uuid, url1, url2 = inputs.sample
-        mock_bam = '.'.join(url1.split('.')[:-2])[:-2] + ".bam"
-        samples = [[uuid, [url1, url2], mock_bam]]
-    else:
-        samples = []
-
-        with open(inputs.manifest, 'r') as f_in:
-            for line in f_in:
-                if not line.isspace() and not line.startswith('#'):
-                    line = line.strip().split('\t')
-                    assert len(line) == 3, 'Improper formatting. Expected UUID\tURl1\tURL2. Received: {}'.format(line)
-                    uuid = line[0]
-                    urls = line[1:]
-                    mock_bam = '.'.join(line[1].split('.')[:-2])[:-2] + ".bam"
-                    samples.append((uuid, urls, mock_bam))
-    inputs.maxCores = int(inputs.maxCores) if hasattr(inputs, "maxCores") else sys.maxint
     inputs.cores = min(inputs.maxCores, multiprocessing.cpu_count())
-    job.fileStore.logToMaster('Parsed configuration file.')
-    job.addChildJobFn(map_job, download_sample, samples, inputs, shared_ids, cores=1, disk=inputs.file_size)
-
-
-def download_sample(job, sample, inputs, ids):
+    shared_ids = {}
+    urls = [('amb', inputs.amb), ('ann', inputs.ann), ('bwt', inputs.bwt),
+            ('pac', inputs.pac), ('sa', inputs.sa)]
+    # If all inputs provided, download from S3
+    download_ref = job.wrapJobFn(download_url_job, inputs.ref, disk='3G')
+    faidx = job.wrapJobFn(run_samtools_faidx, download_ref.rv())
+    # If all the bwa index files are provided
+    if all(urls):
+        if inputs.alt:
+            urls.append(('alt', inputs.alt))
+        else:
+            shared_ids['alt'] = None
+        if inputs.fai:
+            urls.append(('fai', inputs.fai))
+        else:
+            # If just fai is absent, create it
+            job.addChild(download_ref)
+            download_ref.addChild(faidx)
+            shared_ids['ref'] = download_ref.rv()
+            shared_ids['fai'] = faidx.rv()
+        # Download the reference, bwa index files, and fai / alt if provided
+        for name, url in urls:
+            shared_ids[name] = job.addChildJobFn(download_url_job, url).rv()
+    else:
+        # Create bwa index files since they were not provided
+        bwa_index = job.wrapJobFn(run_bwa_index, download_ref.rv())
+        job.addChild(download_ref)
+        download_ref.addChild(faidx)
+        download_ref.addChild(bwa_index)
+        shared_ids['ref'] = download_ref.rv()
+        shared_ids['fai'] = faidx.rv()
+        for x, name in enumerate(['amb', 'ann', 'bwt', 'pac', 'sa']):
+            shared_ids[name] = bwa_index.rv(x)
+        job.addFollowOnJobFn(map_job, download_sample_and_align, samples, inputs, shared_ids)
```

Comment:
Also, can you document that `map_job` maps over the list elements in `samples` or something. That code gave me pause and I had to look across files to suss out what was going on.